### PR TITLE
Validate trust threshold config option

### PR DESCRIPTION
**Pull request by hu55a1n1**
Closes: #1149

## Description

Add validation to the config.toml trust threshold and warn that trust threshold is an advanced feature. 

______

For contributor use:

- [ ] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [ ] If applicable: Unit tests written, added test to CI.
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: guide/src/config.md [hunk 1]
```diff
@@ -101,10 +101,12 @@ A `chains` section includes parameters related to a chain and the full node to w
 
 * __trusting_period__: *(string)* Specify the amount of time to be used as the light client trusting period. It should be significantly less than the unbonding period (e.g. unbonding period = 3 weeks, trusting period = 2 weeks). Default: `14days` (336 hours)
 
-* __trust_threshold__: *(table)* Specify the trust threshold for the light client, ie. the maximum fraction of validators which have changed between two blocks. Default: `{ numerator = '1', denominator = '3' }`, ie. 1/3.
+* __trust_threshold__ <sup>(advanced)</sup>: *(table)* Specify the trust threshold for the light client, ie. the maximum fraction of validators which have changed between two blocks. Default: `{ numerator = '1', denominator = '3' }`, ie. 1/3.
```
Comment (adizere) [inline guide/src/config.md:104]: Cool!

BTW, it is also important to add the warning in the ibc-rs/config.toml file

Comment (hu55a1n1) [reply guide/src/config.md:104]: Done 👍

File: config.toml [hunk 2]
```diff
@@ -98,6 +98,7 @@ trusting_period = '14days'
 # Specify the trust threshold for the light client, ie. the maximum fraction of validators
 # which have changed between two blocks.
 # Default: { numerator = '1', denominator = '3' }, ie. 1/3.
+# Warning: This is an advanced feature! Modify with caution.
 trust_threshold = { numerator = '1', denominator = '3' }
 [chains.filters]
```
Comment (romac) [inline config.toml:103]:
```suggestion
trust_threshold = { numerator = '1', denominator = '3' }

[chains.filters]
```

File: relayer-cli/src/config.rs [hunk 3]
```diff
@@ -26,16 +27,43 @@ pub enum Error {
...
+    /// Invalid trust threshold
+    #[error("config file specifies an invalid trust threshold ({0}) for the chain with id {1}, caused by: {2}")]
+    InvalidTrustThreshold(TrustThreshold, ChainId, String),
 }
 
 /// Method for syntactic validation of the input
 /// configuration file.
 pub fn validate_config(config: &Config) -> Result<(), Error> {
-    // Check for duplicate chain configuration.
+    // Check for duplicate chain configuration and invalid trust-thresholds
```
Comment (romac) [inline relayer-cli/src/config.rs:39]:
```suggestion
    // Check for duplicate chain configuration and invalid trust thresholds
```

File: relayer-cli/src/config.rs [hunk 4]
```diff
@@ -26,16 +27,43 @@ pub enum Error {
...
+                c.trust_threshold,
+                c.id.clone(),
+                "Threshold cannot be < 1/3".to_owned(),
+            ));
+        }
+        if c.trust_threshold.numerator >= c.trust_threshold.denominator {
+            return Err(Error::InvalidTrustThreshold(
+                c.trust_threshold,
+                c.id.clone(),
+                "Threshold cannot be >= 1".to_owned(),
```
Comment (romac) [inline relayer-cli/src/config.rs:65]:
```suggestion
                "threshold cannot be >= 1".to_owned(),
```

File: relayer-cli/src/config.rs [hunk 5]
```diff
@@ -26,16 +27,43 @@ pub enum Error {
...
+                c.trust_threshold,
+                c.id.clone(),
+                "Denominator cannot be zero".to_owned(),
+            ));
+        }
+        if c.trust_threshold.numerator * 3 < c.trust_threshold.denominator {
+            return Err(Error::InvalidTrustThreshold(
+                c.trust_threshold,
+                c.id.clone(),
+                "Threshold cannot be < 1/3".to_owned(),
```
Comment (romac) [inline relayer-cli/src/config.rs:58]:
```suggestion
                "trust threshold cannot be < 1/3".to_owned(),
```

File: relayer-cli/src/config.rs [hunk 6]
```diff
@@ -26,16 +27,43 @@ pub enum Error {
...
+        if !unique_chain_ids.insert(c.id.clone()) {
+            return Err(Error::DuplicateChains(c.id.clone()));
+        }
+
+        // TODO: move below validation to tendermint::trust_threshold
+        if c.trust_threshold.denominator == 0 {
+            return Err(Error::InvalidTrustThreshold(
+                c.trust_threshold,
+                c.id.clone(),
+                "Denominator cannot be zero".to_owned(),
```
Comment (romac) [inline relayer-cli/src/config.rs:51]:
```suggestion
                "trust threshold denominator cannot be zero".to_owned(),
```

File: relayer-cli/src/config.rs [hunk 7]
```diff
@@ -26,16 +27,43 @@ pub enum Error {
...
     let mut unique_chain_ids = BTreeSet::new();
-    for chain_id in config.chains.iter().map(|c| c.id.clone()) {
-        if !unique_chain_ids.insert(chain_id.clone()) {
-            return Err(Error::DuplicateChains(chain_id));
+    for c in &config.chains {
+        if !unique_chain_ids.insert(c.id.clone()) {
+            return Err(Error::DuplicateChains(c.id.clone()));
+        }
+
+        // TODO: move below validation to tendermint::trust_threshold
```
Comment (romac) [inline relayer-cli/src/config.rs:46]: Agreed, would be nice to move this into its own function.

Comment (hu55a1n1) [reply relayer-cli/src/config.rs:46]: Thank you for the suggestions @romac! I created an issue in tendermint-rs (https://github.com/informalsystems/tendermint-rs/issues/924) to move this validation code there. Would you like me to refactor this into a function here as well?

Comment (romac) [reply relayer-cli/src/config.rs:46]: Yes please!

Comment (romac) [reply relayer-cli/src/config.rs:46]: I will do it now :)

**Review by romac — APPROVED** [review]

**Pull request merged**